planemo/autoupdate.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import absolute_import
+
+import re
+import xml.etree.ElementTree as ET
+import planemo.conda
+
+from planemo.io import error, info
+
+def autoupdate(tool_path, dry_run=False):
+    """
+    Autoupdate an XML file
+    """
+    xml_tree = ET.parse(tool_path)
+    requirements = find_requirements(xml_tree)
+
+    for macro_import in requirements['imports']:
+        # recursively check macros
+        macro_requirements = autoupdate('/'.join(tool_path.split('/')[:-1] + [macro_import]), dry_run)
+        for requirement in macro_requirements:
+            if requirement not in requirements:
+                requirements[requirement] = macro_requirements[requirement]
+
+    if not requirements.get('@TOOL_VERSION@'):
+        # if tool_version is not specified, finish without changes
+        error("The @TOOL_VERSION@ token is not specified in {}. This is required for autoupdating.".format(tool_path))
+        return requirements
+    updated_main_req = get_latest_versions({requirements.get('main_req'): requirements.get('@TOOL_VERSION@')})
+    if updated_main_req[requirements.get('main_req')] == requirements.get('@TOOL_VERSION@'):
+        # check main_req is up-to-date; if so, finish without changes
+        info("No updates required or madeto {}.".format(tool_path))
+        return requirements
+
+    if dry_run:
+        error("Update required to {}! Tool main requirement has version {}, newest conda version is {}".format(tool_path, requirements.get('@TOOL_VERSION@'), updated_main_req[requirements.get('main_req')]))
+        return requirements
+
+    # if main_req is not up-to-date, update everything
+    updated_version_dict = get_latest_versions(requirements.get('other_reqs'))
+    update_requirements(tool_path, xml_tree, updated_version_dict, updated_main_req)
+    info("Tool {} updated.".format(tool_path))
+    return requirements
+
+def find_requirements(xml_tree):
+    """
+    Get all requirements with versions as a dictionary of the form
+    {'main_req': main requirement, 'other_reqs': {req1: version, ... },
+        'imports: ['macros.xml'], '*VERSION@': '...'}
+    """
+    requirements = {'other_reqs': {}, 'imports': []}
+
+    # get tokens
+    for token in xml_tree.iter("token"):
+        if 'VERSION@' in token.attrib.get('name', ''):
+            requirements[token.attrib['name']] = token.text
+    for macro_import in xml_tree.iter("import"):
+        requirements['imports'].append(macro_import.text)
+
+    # get requirements
+    for requirement in xml_tree.iter("requirement"):
+        if requirement.attrib.get('version') == '@TOOL_VERSION@':
+            requirements['main_req'] = requirement.text
+        else:
+            requirements['other_reqs'][requirement.text] = requirement.attrib.get('version')
+    return requirements
+
+
+def get_latest_versions(version_dict):
+    """
+    Update a dict with current conda versions for tool requirements
+    """
```

Comment:
This is pretty awesome!

Reply:
Thank you @jmchilton - I think this is ready for a review

Reply:
Actually, I take that back, there are still some things I can fix. Though any overall comments about the implementation would be welcome.